connectivity/service/src/main/java/org/eclipse/ditto/connectivity/service/mapping/CloudEventsMapper.java

```diff
@@ -0,0 +1,166 @@
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.ditto.connectivity.service.mapping;
+
+import static java.util.Collections.singletonList;
+import static org.eclipse.ditto.json.JsonFactory.newObject;
+
+import java.util.Base64;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+
+import org.eclipse.ditto.base.model.exceptions.DittoJsonException;
+import org.eclipse.ditto.base.model.headers.DittoHeaders;
+import org.eclipse.ditto.connectivity.api.ExternalMessage;
+import org.eclipse.ditto.connectivity.api.ExternalMessageFactory;
+import org.eclipse.ditto.connectivity.model.MessageMappingFailedException;
+import org.eclipse.ditto.json.JsonFactory;
+import org.eclipse.ditto.json.JsonObject;
+import org.eclipse.ditto.json.JsonValue;
+import org.eclipse.ditto.protocol.Adaptable;
+import org.eclipse.ditto.protocol.JsonifiableAdaptable;
+import org.eclipse.ditto.protocol.ProtocolFactory;
+
+/**
+ * A message mapper implementation for the Mapping incoming CloudEvents to Ditto Protocol.
+ */
+@PayloadMapper(alias = {"CloudEvents",
+    // legacy full qualified name
+    "org.eclipse.ditto.connectivity.service.mapping.CloudEventsMapper"})
+public final class CloudEventsMapper extends AbstractMessageMapper {
+
+  static final String CE_ID = "ce-id";
+  static final String CE_TYPE = "ce-type";
+  static final String CE_SOURCE = "ce-source";
+  static final String CE_SPECVERSION = "ce-specversion";
+  static final String STRUCTURED_CONTENT_TYPE = "application/cloudevents+json";
+  static final String BINARY_CONTENT_TYPE = "application/vnd.eclipse.ditto+json";
+  static final String SPECVERSION = "specversion";
+  static final String ID = "id";
+  static final String SOURCE = "source";
+  static final String TYPE = "type";
+  static final String OUTBOUNDTYPE = "org.eclipse.ditto.outbound";
+  static final String OUTBOUNDSPECVERSION = "1.0";
+  static final String OUTBOUNDSOURCE = "org.eclipse.ditto";
+  static final String DATA = "data";
+  static final String DATA_BASE64 = "data_base64";
+
+  @Override
+  public List<Adaptable> map(final ExternalMessage message) {
+    final String payload = extractPayloadAsString(message);
+    final String contentType = message.findContentType().orElse("");
+    if (contentType.equals(BINARY_CONTENT_TYPE)) {
+      if (isBinaryCloudEvent(message)) {
+        final JsonifiableAdaptable binaryAdaptable = DittoJsonException.wrapJsonRuntimeException(
+            () -> ProtocolFactory.jsonifiableAdaptableFromJson(newObject(payload)));
+        DittoHeaders headers = binaryAdaptable.getDittoHeaders();
```

Comment:
I think we should extract the `"id"` of the incoming CE and use it as `correlation-id` of the DittoHeaders, e.g.:
```suggestion
        final DittoHeaders headers = binaryAdaptable.getDittoHeaders()
                .toBuilder()
                .correlationId(message.getHeaders().get(CE_ID))
                .build();
```